eth2util/deposit/deposit_data.go

```diff
@@ -18,142 +18,219 @@ package deposit
+	zeroBytes11 = []byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
 
-	// Eth1WithdrawalAddress is the Ethereum withdrawal address.
-	Eth1WithdrawalAddress string
+	// zeroes refers to a zeroed out 32 byte array.
+	zeroBytes32 = []byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
 
-	// DepositMessageRoot is the hash tree root of DepositMessage.
-	DepositMessageRoot eth2p0.Root
+	// DOMAIN_DEPOSIT. See spec: https://benjaminion.xyz/eth2-annotated-spec/phase0/beacon-chain/#domain-types
+	depositDomainType = []byte{0x03, 0x00, 0x00, 0x00}
+)
+
+// GetMessageRoot returns both the hash root and the signing root of the deposit message.
+func GetMessageRoot(pubkey eth2p0.BLSPubKey, withdrawalCreds [32]byte, forkVersion eth2p0.Version) (eth2p0.Root, eth2p0.Root, error) {
+	dm := eth2p0.DepositMessage{
+		PublicKey:             pubkey,
+		WithdrawalCredentials: withdrawalCreds[:],
+		Amount:                eth2p0.Gwei(validatorAmt),
+	}
+
+	hashRoot, err := dm.HashTreeRoot()
+	if err != nil {
+		return eth2p0.Root{}, eth2p0.Root{}, errors.Wrap(err, "deposit message hash root")
+	}
 
-	// Signature is constructed from DepositMessageRoot combined with DOMAIN_DEPOSIT.
-	Signature eth2p0.BLSSignature
+	signingRoot, err := GetSigningRoot(forkVersion, hashRoot)
+	if err != nil {
+		return eth2p0.Root{}, eth2p0.Root{}, errors.Wrap(err, "deposit message root")
+	}
 
-	// ForkVersion identifies the network/chainID.
-	ForkVersion eth2p0.Version
+	return hashRoot, signingRoot, nil
 }
 
-func (d depositData) HashTreeRoot() ([32]byte, error) {
-	b, err := ssz.HashWithDefaultHasher(d)
+// GetDataRoot returns both the hash root and the signing root of the deposit data.
+func GetDataRoot(pubkey eth2p0.BLSPubKey, withdrawalCreds [32]byte, sig eth2p0.BLSSignature, forkVersion eth2p0.Version) (eth2p0.Root, eth2p0.Root, error) {
+	dd := eth2p0.DepositData{
+		PublicKey:             pubkey,
+		WithdrawalCredentials: withdrawalCreds[:],
+		Amount:                eth2p0.Gwei(validatorAmt),
+		Signature:             sig,
+	}
+
+	hashRoot, err := dd.HashTreeRoot()
 	if err != nil {
-		return [32]byte{}, errors.Wrap(err, "hash deposit data")
+		return eth2p0.Root{}, eth2p0.Root{}, errors.Wrap(err, "deposit data hash root")
 	}
 
-	return b, nil
+	// calculate depositData root
+	signingRoot, err := GetSigningRoot(forkVersion, hashRoot)
+	if err != nil {
+		return eth2p0.Root{}, eth2p0.Root{}, errors.Wrap(err, "deposit data root")
+	}
+
+	return hashRoot, signingRoot, nil
 }
 
-func (d depositData) HashTreeRootWith(hh *ssz.Hasher) error {
-	idx := hh.Index()
+// NewDepositData returns the json serialized DepositData.
+func NewDepositData(pubkey eth2p0.BLSPubKey, withdrawalAddr string, sig eth2p0.BLSSignature, network string) ([]byte, error) {
+	forkVersion := NetworkToForkVersion(network)
```

Comment:
I would unexport all functions, except `MarshalDepositData` and `GetMessageRoot`

Reply:
rename `NewDepositData` to `MarshalDepositData`

Reply:
Yes, I agree `New` should be reserved for constructors.

let's as you say unexport as default, and export only things that we need outside